api/src/main/java/org/apache/iceberg/transforms/Timestamps.java

```diff
@@ -31,54 +33,160 @@
+      case "year":
+        return get(type, ChronoUnit.YEARS);
+      case "month":
+        return get(type, ChronoUnit.MONTHS);
+      case "day":
+        return get(type, ChronoUnit.DAYS);
+      case "hour":
+        return get(type, ChronoUnit.HOURS);
+      default:
+        throw new IllegalArgumentException(
+            "Unsupported source/result type units: " + type + " -> " + resultTypeUnit);
+    }
+  }
+
+  static Timestamps get(Types.TimestampNanoType type, String resultTypeUnit) {
+    switch (resultTypeUnit.toLowerCase(Locale.ENGLISH)) {
+      case "year":
+        return get(type, ChronoUnit.YEARS);
+      case "month":
+        return get(type, ChronoUnit.MONTHS);
+      case "day":
+        return get(type, ChronoUnit.DAYS);
+      case "hour":
+        return get(type, ChronoUnit.HOURS);
+      default:
+        throw new IllegalArgumentException(
+            "Unsupported source/result type units: " + type + " -> " + resultTypeUnit);
+    }
+  }
+
+  static Timestamps get(Types.TimestampType type, ChronoUnit resultTypeUnit) {
+    switch (resultTypeUnit) {
+      case YEARS:
+        return YEAR_FROM_MICROS;
+      case MONTHS:
+        return MONTH_FROM_MICROS;
+      case DAYS:
+        return DAY_FROM_MICROS;
+      case HOURS:
+        return HOUR_FROM_MICROS;
+      default:
+        throw new IllegalArgumentException(
+            "Unsupported source/result type units: " + type + " -> " + resultTypeUnit);
+    }
+  }
+
+  static Timestamps get(Types.TimestampNanoType type, ChronoUnit resultTypeUnit) {
+    switch (resultTypeUnit) {
+      case YEARS:
+        return YEAR_FROM_NANOS;
+      case MONTHS:
+        return MONTH_FROM_NANOS;
+      case DAYS:
+        return DAY_FROM_NANOS;
+      case HOURS:
+        return HOUR_FROM_NANOS;
+      default:
+        throw new IllegalArgumentException(
+            "Unsupported source/result type units: " + type + " -> " + resultTypeUnit);
+    }
+  }
+
+  enum ResultTypeUnit {
+    YEARS(ChronoUnit.YEARS),
+    MONTHS(ChronoUnit.MONTHS),
+    DAYS(ChronoUnit.DAYS),
+    HOURS(ChronoUnit.HOURS),
+    MICROS(ChronoUnit.MICROS),
+    NANOS(ChronoUnit.NANOS),
+    ;
+
+    private final ChronoUnit unit;
+
+    ResultTypeUnit(final ChronoUnit unit) {
+      this.unit = unit;
+    }
+
+    Duration duration() {
+      return unit.getDuration();
+    }
+  }
 
   @Immutable
   static class Apply implements SerializableFunction<Long, Integer> {
-    private final ChronoUnit granularity;
+    private final ChronoUnit sourceTypeUnit;
+    private final ResultTypeUnit resultTypeUnit;
 
-    Apply(ChronoUnit granularity) {
-      this.granularity = granularity;
+    Apply(ChronoUnit sourceTypeUnit, ResultTypeUnit resultTypeUnit) {
+      this.sourceTypeUnit = sourceTypeUnit;
+      this.resultTypeUnit = resultTypeUnit;
     }
 
     @Override
-    public Integer apply(Long timestampMicros) {
-      if (timestampMicros == null) {
+    public Integer apply(Long timestampUnits) {
```

Comment:
Feels a bit odd to call this variable `timestampUnits`. The actual units is passed to the class as part of `sourceTypeUnit`. I'd just call it `timetamp`.

Reply:
done